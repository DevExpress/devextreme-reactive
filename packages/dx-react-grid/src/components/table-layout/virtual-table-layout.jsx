import * as React from 'react';
import { findDOMNode } from 'react-dom';
import * as PropTypes from 'prop-types';
import { ColumnGroup } from './column-group';
import { RowLayout } from './row-layout';
import {
  getVisibleRows,
  firstVisibleRowOffset,
} from './virtual-table-utils';

export class VirtualTableLayout extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      rowHeights: new Map(),
      viewportTop: 0,
    };

    this.rowRefs = new Map();
    this.updateViewport = this.updateViewport.bind(this);
    this.registerRowRef = this.registerRowRef.bind(this);
    this.getRowHeight = this.getRowHeight.bind(this);
  }
  componentDidMount() {
    this.storeRowHeights();
  }
  componentWillReceiveProps(nextProps) {
    if (
      this.props.headerRows !== nextProps.headerRows ||
      this.props.rows !== nextProps.rows
    ) {
      const { rowHeights: prevRowHeight } = this.state;
      this.setState({
        rowHeights: [...nextProps.headerRows, ...nextProps.rows].reduce(
          (acc, row) => {
            const rowHeight = prevRowHeight.get(row.key);
            if (rowHeight !== undefined) {
              acc.set(row.key, rowHeight);
            }
            return acc;
          },
          new Map(),
        ),
      });
    }
  }
  componentDidUpdate() {
    this.storeRowHeights();
  }
  getRowHeight(row) {
    const storedHeight = this.state.rowHeights.get(row.key);
    if (storedHeight !== undefined) return storedHeight;
    if (row.height) return row.height;
    return this.props.estimatedRowHeight;
  }
  getVisibleRows({ rows, headerRows, height } = this.props, top = this.state.viewportTop) {
    const headHeight = headerRows.reduce((acc, row) => acc + this.getRowHeight(row), 0);
    return getVisibleRows(rows, top, height - headHeight, this.getRowHeight);
  }
  storeRowHeights() {
    const rowsWithChangedHeights = Array.from(this.rowRefs.entries())
      // eslint-disable-next-line react/no-find-dom-node
      .map(([row, ref]) => [row, findDOMNode(ref).getBoundingClientRect().height])
      .filter(([row, height]) => height !== this.getRowHeight(row));

    if (rowsWithChangedHeights.length) {
      const prevVisibleBodyRows = this.getVisibleRows();

      const { rowHeights } = this.state;
      rowsWithChangedHeights
        .forEach(([row, height]) => rowHeights.set(row.key, height));

      this.setState({
        rowHeights,
      });

      const visibleBodyRows = this.getVisibleRows();
      const scrollOffset = firstVisibleRowOffset(prevVisibleBodyRows, visibleBodyRows);
      if (scrollOffset !== 0) {
        // eslint-disable-next-line react/no-find-dom-node
        findDOMNode(this).scrollTop += scrollOffset;
      }
    }
  }
  registerRowRef(row, ref) {
    if (ref === null) {
      this.rowRefs.delete(row);
    } else {
      this.rowRefs.set(row, ref);
    }
  }
  updateViewport(e) {
    const node = e.target;

    if (node !== e.currentTarget) {
      return;
    }

    // NOTE: prevent iOS to flicker in bounces
    if (node.scrollTop < 0 ||
      node.scrollLeft < 0 ||
      node.scrollLeft + node.clientWidth > node.scrollWidth ||
      node.scrollTop + node.clientHeight > node.scrollHeight) {
      return;
    }

    if (this.viewportTop !== node.scrollTop) {
      this.setState({
        viewportTop: node.scrollTop,
      });
    }
  }
  render() {
    const {
      headerRows, columns,
      minWidth, height,
      containerComponent: Container,
      headTableComponent: HeadTable,
      tableComponent: Table,
      headComponent: Head,
      bodyComponent: Body,
      rowComponent, cellComponent,
    } = this.props;
    const visibleBodyRows = this.getVisibleRows();

    return (
      <Container
        style={{ height: `${height}px` }}
        onScroll={this.updateViewport}
      >
        {!!headerRows.length && (
          <HeadTable
            style={{ minWidth: `${minWidth}px` }}
          >
            <ColumnGroup columns={columns} />
            <Head>
              {headerRows.map(row => (
                <RowLayout
                  key={row.key}
                  ref={ref => this.registerRowRef(row, ref)}
                  row={row}
                  columns={columns}
                  rowComponent={rowComponent}
                  cellComponent={cellComponent}
                />
              ))}
            </Head>
          </HeadTable>
        )}
        <Table
          style={{ minWidth: `${minWidth}px` }}
        >
          <ColumnGroup columns={columns} />
          <Body>
            {visibleBodyRows.map((visibleRow) => {
              if (visibleRow.type === 'stub') {
                return (
                  <tr key={visibleRow.key} style={{ height: `${visibleRow.height}px` }} />
                );
              }
              const { row } = visibleRow;
              return (
                <RowLayout
                  key={row.key}
                  ref={ref => this.registerRowRef(row, ref)}
                  row={row}
                  columns={columns}
                  rowComponent={rowComponent}
                  cellComponent={cellComponent}
                />
              );
            })}
          </Body>
        </Table>
      </Container>
    );
  }
}

VirtualTableLayout.propTypes = {
  minWidth: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  headerRows: PropTypes.array,
  rows: PropTypes.array.isRequired,
  columns: PropTypes.array.isRequired,
  cellComponent: PropTypes.func.isRequired,
  rowComponent: PropTypes.func.isRequired,
  bodyComponent: PropTypes.func.isRequired,
  headComponent: PropTypes.func,
  tableComponent: PropTypes.func.isRequired,
  headTableComponent: PropTypes.func,
  containerComponent: PropTypes.func.isRequired,
  estimatedRowHeight: PropTypes.number.isRequired,
};

VirtualTableLayout.defaultProps = {
  headerRows: [],
  headComponent: () => null,
  headTableComponent: () => null,
};
