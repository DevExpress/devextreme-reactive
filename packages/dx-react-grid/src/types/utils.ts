/** Describes a row cache */
export interface RowCache {
  /** Returns `take` number of rows starting with the `skip` row from the cache */
  getRows(skip: number, take: number): any[];
  /** Adds `rows` to the cache */
  setRows(skip: number, rows: ReadonlyArray<any>): void;
  /** Clears the cache */
  invalidate(): void;
}
