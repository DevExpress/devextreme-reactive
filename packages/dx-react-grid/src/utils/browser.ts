import { extend } from './extend';

const webkitRegExp = /(webkit)[ /]([\w.]+)/;
const ieRegExp = /(msie) (\d{1,2}\.\d)/;
const ie11RegExp = /(trident).*rv:(\d{1,2}\.\d)/;
const msEdge = /(edge)\/((\d+)?[\w.]+)/;
const mozillaRegExp = /(mozilla)(?:.*? rv:([\w.]+))/;

interface Result {
  webkit?: boolean | undefined;
  version?: string | null;
}

const browserFromUA = (uaText) => {
  const ua = uaText.toLowerCase();

  const result: Result = { };
  const matches =
            ieRegExp.exec(ua) ||
            ie11RegExp.exec(ua) ||
            msEdge.exec(ua) ||
            ua.indexOf('compatible') < 0 && mozillaRegExp.exec(ua) ||
            webkitRegExp.exec(ua) ||
            [];
  let browserName = matches[1];
  let browserVersion: RegExpExecArray | null | string = matches[2];

  if (browserName === 'webkit') {
    result.webkit = true;

    if (ua.indexOf('chrome') >= 0 || ua.indexOf('crios') >= 0) {
      browserName = 'chrome';
      browserVersion = /(?:chrome|crios)\/(\d+\.\d+)/.exec(ua);
      browserVersion = browserVersion && browserVersion[1];
    } else if (ua.indexOf('fxios') >= 0) {
      browserName = 'mozilla';
      browserVersion = /fxios\/(\d+\.\d+)/.exec(ua);
      browserVersion = browserVersion && browserVersion[1];
    } else if (ua.indexOf('safari') >= 0 && /version|phantomjs/.test(ua)) {
      browserName = 'safari';
      browserVersion = /(?:version|phantomjs)\/([0-9.]+)/.exec(ua);
      browserVersion = browserVersion && browserVersion[1];
    } else {
      browserName = 'unknown';
      browserVersion = /applewebkit\/([0-9.]+)/.exec(ua);
      browserVersion = browserVersion && browserVersion[1];
    }
  }

  if (browserName === 'trident' || browserName === 'edge') {
    browserName = 'msie';
  }

  if (browserName) {
    result[browserName] = true;
    result.version = browserVersion;
  }

  return result;
};
export const browser = extend({ _fromUA: browserFromUA }, browserFromUA(navigator.userAgent));
