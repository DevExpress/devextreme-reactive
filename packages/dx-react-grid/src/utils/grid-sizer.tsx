import * as React from 'react';
import { Sizer, SizerProps } from '@devexpress/dx-react-core';
import { TABLE_STUB_TYPE, CollapsedGrid } from '@devexpress/dx-grid-core';
import { ReadonlyObject } from '@devexpress/dx-core';

// tslint:disable-next-line: max-line-length
export class GridSizer extends React.PureComponent<SizerProps & {collapsedGrid: ReadonlyObject<CollapsedGrid>}> {
  ref = React.createRef<Sizer>();

  componentDidUpdate() {
    const { collapsedGrid: { rows, columns } } = this.props;

    const hasStubColumn = columns.some(col => col.type === TABLE_STUB_TYPE);
    const hasStubRow = rows.some(row => row.type === TABLE_STUB_TYPE);

    if (hasStubColumn || hasStubRow) {
      this.ref.current!.setupListeners();
    }
  }

  render() {
    const { collapsedGrid, ...restProps } = this.props;
    return <Sizer ref={this.ref} {...restProps} />;
  }
}
