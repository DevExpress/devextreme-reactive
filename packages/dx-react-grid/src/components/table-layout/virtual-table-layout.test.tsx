import * as React from 'react';
import { findDOMNode } from 'react-dom';
import { shallow, mount } from 'enzyme';
import { Sizer } from '@devexpress/dx-react-core';
import { GridSizer } from '../../utils/grid-sizer';
import {
  getCollapsedGrids,
  TABLE_FLEX_TYPE,
  emptyViewport,
} from '@devexpress/dx-grid-core';
import { setupConsole } from '@devexpress/dx-testing';
import { VirtualTableLayout } from './virtual-table-layout';

jest.mock('react-dom', () => ({
  findDOMNode: jest.fn(),
}));
jest.mock('@devexpress/dx-grid-core', () => {
  const actual = require.requireActual('@devexpress/dx-grid-core');
  jest.spyOn(actual, 'getCollapsedGrids');
  jest.spyOn(actual, 'getColumnWidthGetter');
  return actual;
});
jest.mock('./column-group', () => ({
  ColumnGroup: () => null,
}));
jest.mock('@devexpress/dx-react-core', () => {
  const { Component } = require.requireActual('react');
  return {
    ...require.requireActual('@devexpress/dx-react-core'),
    Sizer: class extends Component {
      componentDidMount() {
        // eslint-disable-next-line react/prop-types
        const { onSizeChange } = this.props;
        onSizeChange({ width: 400, height: 120 });
      }

      render() {
        const {
          containerComponent: Container,
          onSizeChange,
          scrollTop,
          ...restProps
        } = this.props;
        return (
          <Container {...restProps} />
        );
      }
    },
    // tslint:disable-next-line: max-classes-per-file
    RefHolder: class extends Component {
      render() {
        // eslint-disable-next-line react/prop-types
        const { children: propsChildren } = this.props;
        return propsChildren;
      }
    },
  };
});
jest.mock('../../utils/grid-sizer', () => {
  const { Component } = require.requireActual('react');
  return {
    ...require.requireActual('../../utils/grid-sizer'),
    // tslint:disable-next-line: max-classes-per-file
    GridSizer: class extends Component {
      render() {
        const { collapsedGrid, ...restProps } = this.props;
        return (
          <Sizer {...restProps} />
        );
      }
    },
  };
});

// tslint:disable-next-line: max-classes-per-file
class VirtualTableLayoutWrapper extends React.Component<any, any> {
  state = {
    viewport: emptyViewport,
  };
  setViewport = viewport => this.setState({ viewport });

  render() {
    const { viewport } = this.state;
    return (
      <VirtualTableLayout
        {...this.props}
        setViewport={this.setViewport}
        viewport={viewport}
      />
    );
  }
}

const defaultProps = {
  columns: [
    { key: 'a', column: { name: 'a' } },
    { key: 'b', column: { name: 'b' } },
    { key: 'c', column: { name: 'c' } },
    { key: 'd', column: { name: 'd' } },
    { key: 'e', column: { name: 'e' } },
  ],
  minWidth: 400,
  minColumnWidth: 120,
  height: 100,
  estimatedRowHeight: 40,
  bodyRows: [
    { key: 1 },
    { key: 2 },
    { key: 3 },
    { key: 4 },
    { key: 5 },
    { key: 6 },
    { key: 7 },
    { key: 8 },
    { key: 9 },
  ],
  viewport: {
    top: 0,
    left: 0,
    width: 400,
    height: 120,
    columns: [[0, 4]],
    rows: [0, 5],
    headerRows: [0, 0],
    footerRows: [0, 0],
  },
  setViewport: jest.fn(),
  loadedRowsStart: 0,
  totalRowCount: 9,
  containerComponent: props => <div {...props} />,
  headTableComponent: ({ tableRef, ...props }) => <table {...props} />,
  tableComponent: ({ tableRef, ...props }) => <table {...props} />,
  headComponent: props => <thead {...props} />,
  bodyComponent: props => <tbody {...props} />,
  rowComponent: () => null,
  cellComponent: () => null,
  getCellColSpan: () => 1,
  tableRef: React.createRef<HTMLTableElement>(),
};

describe('VirtualTableLayout', () => {
  let resetConsole;
  beforeEach(() => {
    resetConsole = setupConsole();
    findDOMNode.mockImplementation(() => ({
      getBoundingClientRect: () => ({
        height: defaultProps.estimatedRowHeight,
      }),
    }));
  });

  afterEach(() => {
    resetConsole();
    jest.clearAllMocks();
  });

  const simulateScroll = (tree, props) => {
    const target = {
      scrollTop: 0,
      scrollLeft: 0,
      ...props,
    };

    const eventData = {
      target,
      currentTarget: target,
    };

    tree
      .find(Sizer)
      .find(defaultProps.containerComponent)
      .prop('onScroll')(eventData);
    tree.update();
  };

  it('should render correct layout', () => {
    const tree = shallow((
      <VirtualTableLayout
        {...defaultProps}
        headerRows={defaultProps.bodyRows.slice(0, 1)}
        footerRows={defaultProps.bodyRows.slice(0, 1)}
      />
    ));

    expect(tree.find(GridSizer).dive().dive())
      .toMatchSnapshot();
  });

  it('should provide scrollTop property', () => {
    const scrollTop = 100;
    const tree = shallow((
      <VirtualTableLayout
        {...defaultProps}
        headerRows={defaultProps.bodyRows.slice(0, 1)}
        footerRows={defaultProps.bodyRows.slice(0, 1)}
        scrollTop={scrollTop}
      />
    ));

    expect(tree.find(GridSizer).prop('scrollTop'))
      .toEqual(scrollTop);
  });

  it('should not render width for a flex column', () => {
    const columns = [
      { key: 'col0', width: 100 },
      { key: 'col1', width: 100 },
      { key: 'col_flex', type: TABLE_FLEX_TYPE },
    ];
    const rows = [{ key: 0 }];

    getCollapsedGrids
      .mockImplementationOnce((args) => {
        const result = require.requireActual('@devexpress/dx-grid-core').getCollapsedGrids(args);

        expect(result.bodyGrid.columns.find(col => col.key === 'col_flex').width)
          .toBe(null);

        return result;
      });

    mount((
      <VirtualTableLayout
        {...defaultProps}
        headerRows={rows}
        columns={columns}
        viewport={{
          ...defaultProps.viewport,
          columns: [[0, 2]],
        }}
      />
    ));
  });

  describe('viewport', () => {
    it('should pass correct viewport at startup', () => {
      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
          headerRows={defaultProps.bodyRows.slice(0, 2)}
          footerRows={defaultProps.bodyRows.slice(0, 2)}
        />
      ));

      expect(tree.find(defaultProps.containerComponent).props().style)
        .toMatchObject({ height: defaultProps.height });

      expect(getCollapsedGrids).toBeCalledTimes(2);
      expect(getCollapsedGrids.mock.calls[getCollapsedGrids.mock.calls.length - 1][0])
        .toMatchObject({
          viewportLeft: 0,
          containerWidth: 400,
          viewport: {
            columns: [[0, 4]],
            footerRows: [0, 0],
            headerRows: [0, 0],
            rows: [0, 5],
          },
        });
    });

    it('should pass correct viewport at startup when height is auto', () => {
      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
          headerRows={defaultProps.bodyRows.slice(0, 1)}
          height="auto"
        />
      ));

      expect(tree.find(defaultProps.containerComponent).props().style)
        .not.toMatchObject({ height: `${defaultProps.height}px` });
    });

    it('should recalculate viewport on scroll', () => {
      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
          headerRows={defaultProps.bodyRows.slice(0, 1)}
        />
      ));

      simulateScroll(tree, { scrollTop: 100, scrollLeft: 250 });

      const setViewportMock = defaultProps.setViewport.mock;
      expect(setViewportMock.calls[setViewportMock.calls.length - 1][0])
        .toMatchObject({
          top: 100,
          left: 250,
          columns: [[1, 4]],
          footerRows: [0, 0],
          headerRows: [0, 0],
          rows: [2, 4],
        });
    });

    it('should update viewport if column count changed', () => {
      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
        />
      ));
      const setViewportMock = defaultProps.setViewport.mock;
      const initialCallCount = setViewportMock.calls.length;

      tree.setProps({ columns: defaultProps.columns.slice(0, 3) });

      expect(setViewportMock.calls.length)
        .toBeGreaterThan(initialCallCount);
    });

    it('should not update viewport if it is not changed', () => {
      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
          bodyRows={defaultProps.bodyRows.slice(0, 4)}
          viewport={{
            ...defaultProps.viewport,
            rows: [0, 2],
          }}
        />
      ));

      tree.setProps({ bodyRows: defaultProps.bodyRows.slice(4, 9) });

      expect(defaultProps.setViewport).not.toHaveBeenCalled();
    });

    describe('scroll bounce', () => {
      const assertRerenderOnBounce = (shouldRerender, scrollArgs) => {
        const tree = mount((
          <VirtualTableLayoutWrapper
            {...defaultProps}
            headerRows={defaultProps.bodyRows.slice(0, 1)}
          />
        ));
        const initialCallsCount = getCollapsedGrids.mock.calls.length;

        simulateScroll(tree, scrollArgs);

        if (shouldRerender) {
          expect(getCollapsedGrids.mock.calls.length).toBeGreaterThan(initialCallsCount);
        } else {
          expect(getCollapsedGrids.mock.calls.length).toBe(initialCallsCount);
        }
      };

      it('should not re-render on horizontal scroll bounce', () => {
        assertRerenderOnBounce(false, {
          scrollLeft: 200,
          clientWidth: 400,
          scrollWidth: 500,
          scrollTop: 0,
        });
      });

      it('should not re-render on vertical scroll bounce', () => {
        assertRerenderOnBounce(false, {
          scrollTop: 200,
          clientHeight: 400,
          scrollHeight: 500,
          scrollLeft: 0,
        });
      });

      it('should normalize scroll position', () => {
        assertRerenderOnBounce(true, {
          scrollLeft: 201,
          clientWidth: 200,
          scrollWidth: 400,
          scrollTop: 0,
        });
      });

      it('should normalize scroll position by 1px only', () => {
        assertRerenderOnBounce(false, {
          scrollLeft: 202,
          clientWidth: 200,
          scrollWidth: 400,
          scrollTop: 0,
        });
      });
    });
  });

  describe('row heights', () => {
    it('should specify correct row height at startup', () => {
      expect.hasAssertions();

      const rows = [
        { key: 1 },
        { key: 2, height: 10 },
      ];

      getCollapsedGrids
        .mockImplementationOnce((args) => {
          const { getRowHeight } = args;
          expect(getRowHeight(rows[0]))
            .toEqual(defaultProps.estimatedRowHeight);
          expect(getRowHeight(rows[1]))
            .toEqual(10);

          return require.requireActual('@devexpress/dx-grid-core').getCollapsedGrids(args);
        });

      mount((
        <VirtualTableLayout
          {...defaultProps}
          bodyRows={rows}
        />
      ));
    });

    it('should store row height when rendered', () => {
      const rows = [
        { key: 1 },
        { key: 2, height: 10 },
      ];

      findDOMNode.mockImplementation(() => ({
        getBoundingClientRect: () => ({
          height: 50,
        }),
      }));

      mount((
        <VirtualTableLayout
          {...defaultProps}
          bodyRows={rows}
        />
      ));

      const { getRowHeight } = getCollapsedGrids.mock.calls[0][0];
      expect(getRowHeight(rows[0]))
        .toEqual(50);
      expect(getRowHeight(rows[1]))
        .toEqual(50);
    });

    it('should clear row height when rows updated', () => {
      const rows = [
        { key: 11 },
        { key: 12 },
      ];

      findDOMNode.mockImplementation(() => ({
        getBoundingClientRect: () => ({
          height: 50,
        }),
      }));

      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
          bodyRows={rows.slice(0, 2)}
        />
      ));
      tree.setProps({ bodyRows: [rows[0]] });

      const { getRowHeight } = getCollapsedGrids.mock.calls[0][0];
      expect(getRowHeight(rows[1]))
        .toEqual(defaultProps.estimatedRowHeight);
    });
  });
});
