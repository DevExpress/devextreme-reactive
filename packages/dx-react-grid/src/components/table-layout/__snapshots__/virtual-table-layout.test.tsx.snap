// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualTableLayout should render correct layout 1`] = `
<containerComponent
  onScroll={[Function]}
  style={
    Object {
      "height": 100,
    }
  }
>
  <VirtualTableLayoutBlock
    blockRefsHandler={[Function]}
    bodyComponent={[Function]}
    cellComponent={[Function]}
    collapsedGrid={
      Object {
        "columns": Array [
          Object {
            "column": Object {
              "name": "a",
            },
            "key": "a",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "b",
            },
            "key": "b",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "c",
            },
            "key": "c",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "d",
            },
            "key": "d",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "e",
            },
            "key": "e",
            "width": 160,
          },
        ],
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 1,
            },
          },
        ],
      }
    }
    minColumnWidth={120}
    name="header"
    rowComponent={[Function]}
    rowRefsHandler={[Function]}
    tableComponent={[Function]}
    tableRef={
      Object {
        "current": null,
      }
    }
  />
  <VirtualTableLayoutBlock
    blockRefsHandler={[Function]}
    bodyComponent={[Function]}
    cellComponent={[Function]}
    collapsedGrid={
      Object {
        "columns": Array [
          Object {
            "column": Object {
              "name": "a",
            },
            "key": "a",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "b",
            },
            "key": "b",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "c",
            },
            "key": "c",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "d",
            },
            "key": "d",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "e",
            },
            "key": "e",
            "width": 160,
          },
        ],
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 1,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 2,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 3,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 4,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 5,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 6,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 7,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 8,
            },
          },
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 9,
            },
          },
        ],
      }
    }
    marginBottom={600}
    minColumnWidth={120}
    name="body"
    rowComponent={[Function]}
    rowRefsHandler={[Function]}
    tableComponent={[Function]}
    tableRef={
      Object {
        "current": null,
      }
    }
  />
  <VirtualTableLayoutBlock
    blockRefsHandler={[Function]}
    bodyComponent={[Function]}
    cellComponent={[Function]}
    collapsedGrid={
      Object {
        "columns": Array [
          Object {
            "column": Object {
              "name": "a",
            },
            "key": "a",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "b",
            },
            "key": "b",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "c",
            },
            "key": "c",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "d",
            },
            "key": "d",
            "width": 160,
          },
          Object {
            "column": Object {
              "name": "e",
            },
            "key": "e",
            "width": 160,
          },
        ],
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "a",
                  },
                  "key": "a",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "b",
                  },
                  "key": "b",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "c",
                  },
                  "key": "c",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "d",
                  },
                  "key": "d",
                },
              },
              Object {
                "colSpan": 1,
                "column": Object {
                  "column": Object {
                    "name": "e",
                  },
                  "key": "e",
                },
              },
            ],
            "row": Object {
              "key": 1,
            },
          },
        ],
      }
    }
    minColumnWidth={120}
    name="footer"
    rowComponent={[Function]}
    rowRefsHandler={[Function]}
    tableComponent={[Function]}
    tableRef={
      Object {
        "current": null,
      }
    }
  />
</containerComponent>
`;
